# Asssingment 03
## Question 05- Support Vector Machine with Gradient
## Index Number - 17000475
### Youtube Link -

## SVM Implementation

In [93]:
"""
alpha = learing rate
C = reggularization strength
n_iter = number of iterations
"""
def svm_algo(X,Y,alpha, n_iter,C):
    w = np.zeros(X.shape[1]) 
    j=0
    prev_cost = float("inf")
    for n in range(n_iter):
        for i in range(X.shape[0]):
            dw = calculate_cost_gradient(w ,X[i], Y[i],C)
            w = w - alpha*dw

         # stop when algo no longer ocnverge
        if n == 2 ** j or n == n_iter - 1:
            cost = cost_function(w,X,Y,C)
            print("Iteration is:{} and Cost is: {}".format(n, cost))
            # stopping condition
            if abs(prev_cost - cost) < 0.01 * prev_cost:  
                return w
            prev_cost = cost
            j+=1
    return w


In [94]:
def cost_function(w,X,Y,C):        # f(x)
    
    distances = 1 - Y * (np.dot(X, w))
    distances[distances < 0] = 0  
    #hinge_loss = reg_strength * (np.sum(max(0,1 - Y * (np.dot(X, w)))) / X.shape[0])
    hinge_loss = C * (np.sum(distances) / X.shape[0])
    # calculate cost
    cost = 1 / 2 * np.dot(w, w) + hinge_loss
    return cost

In [95]:
def calculate_cost_gradient(w,x,y,C):
    # one example passed
    #print('x',x.shape)
    #print('y',y.shape)
    if type(y) == np.float64 :
        y = np.array([y])
        x = np.array([x])
    distance = 1- (y*np.dot(x,w))
    dw = np.zeros(len(w))
    #print(distance)
    #for j in range(distance.shape[0]):
    if max(0,distance) == 0 :     # hinge loss condition
        di = w
    else:
        di = w-(C*y*x)
    dw = dw+ di
    #dw = dw/len(y)
    return dw


## Breast Cancer Dataset - Preprocessing

In [96]:
import numpy as np 
import pandas as pd 

In [97]:
df = pd.read_csv('datafile.csv',header = None)
df.head()

0   1   2   3   4   5   6   7   8   9   10
0  1000025   5   1   1   1   2   1   3   1   1   2
1  1002945   5   4   4   5   7  10   3   2   1   2
2  1015425   3   1   1   1   2   2   3   1   1   2
3  1016277   6   8   8   1   3   4   3   7   1   2
4  1017023   4   1   1   3   2   1   3   1   1   2

In [98]:
 #replace ? mark  with 0
 df = df.replace(["?"],0)
 #check whether it's removed
df[df.eq("?").any(1)]

Empty DataFrame
Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Index: []

In [99]:
# drop id column
df.drop([0],axis=1,inplace = True)
# replace class column 10 th column 2, 4 value with -1 and, +1
df[10] = df[10].map({2:-1,4:+1})

In [100]:
X = df[:].drop(10,axis=1).to_numpy(dtype='int')
print(X)
print(X.shape)
Y = df[:].pop(10).to_numpy(dtype='int')
print(Y)
print(Y.shape)

[[ 5  1  1 ...  3  1  1]
 [ 5  4  4 ...  3  2  1]
 [ 3  1  1 ...  3  1  1]
 ...
 [ 5 10 10 ...  8 10  2]
 [ 4  8  6 ... 10  6  1]
 [ 4  8  8 ... 10  4  1]]
(699, 9)
[-1 -1 -1 -1 -1  1 -1 -1 -1 -1 -1 -1  1 -1  1  1 -1 -1  1 -1  1  1 -1  1
 -1  1 -1 -1 -1 -1 -1 -1  1 -1 -1 -1  1 -1  1  1 -1  1  1  1  1 -1  1 -1
 -1  1  1  1  1  1  1  1  1  1  1  1  1 -1  1  1 -1  1 -1  1  1 -1 -1  1
 -1  1  1 -1 -1 -1 -1 -1 -1 -1 -1 -1  1  1  1  1 -1 -1 -1 -1 -1 -1 -1 -1
 -1 -1  1  1  1  1 -1  1  1  1  1  1 -1  1 -1  1  1  1 -1 -1 -1  1 -1 -1
 -1 -1  1  1  1 -1  1 -1  1 -1 -1 -1  1 -1 -1 -1 -1 -1 -1 -1 -1 -1  1 -1
 -1 -1  1 -1 -1  1 -1  1  1 -1 -1  1 -1 -1 -1  1  1 -1 -1 -1 -1 -1  1  1
 -1 -1 -1 -1 -1  1  1  1 -1  1 -1  1 -1 -1 -1  1  1 -1  1  1  1 -1  1  1
 -1 -1 -1 -1 -1 -1 -1 -1  1  1 -1 -1 -1  1  1 -1 -1 -1  1  1 -1  1  1  1
 -1 -1  1 -1 -1  1  1  1  1 -1  1  1 -1  1  1  1 -1  1 -1 -1  1  1  1  1
 -1 -1 -1 -1 -1 -1  1  1 -1 -1 -1  1 -1  1  1  1 -1 -1 -1 -1  1  1  1  1
  1 -1  1  1  1 -1  1 -1  1  1 -

In [101]:
# add column of 1 as b value intercept as refering to tutorial
n = X.shape[0]
X= np.hstack((X,np.ones((n,1))))
X.shape

(699, 10)

## Train Test Split

In [102]:
# Train test split
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.33,random_state=200)

## Prediction

In [103]:
# Training 
w = svm_algo(X_train,Y_train,0.000001,100,10000)
# Compute accuracy of training set
Y_train_pred = np.sign(np.dot(X_train,w))
count_train = np.count_nonzero(Y_train_pred + Y_train) # correct count

print('Train accuracy: {}'.format(float(count_train) / X_train.shape[0]))

# Compute accuracy of test set
Y_test_pred = np.sign(np.dot(X_test,w))
count_test = np.count_nonzero(Y_test_pred + Y_test)   # correct count

print('Test accuracy: {}'.format( float(count_test) / X_test.shape[0]))

Iteration is:1 and Cost is: 2395.940328953253
Iteration is:2 and Cost is: 2194.2163800872886
Iteration is:4 and Cost is: 1420.9915607167839
Iteration is:8 and Cost is: 1025.1164487048768
Iteration is:16 and Cost is: 824.635722493646
Iteration is:32 and Cost is: 741.6394743616029
Iteration is:64 and Cost is: 616.8387736086553
Iteration is:99 and Cost is: 564.4976291413645
Train accuracy: 0.9829059829059829
Test accuracy: 0.9567099567099567


## Conclusion
Support vector Machine gave good accuracy than passive aggressive classifier used in Quesion 1. Here I got 98.3% accuracy for training set and 95.6% accuracy for testing set. Considerably a good result for test accuracy.